### PR TITLE
[TE] Rootcause POC - heatmap component, event layouting, metric baselines

### PR DESCRIPTION
**Pull request by apucher**
This PR adds additional functionality to the rootcause frontend POC.

* heatmap component skeleton (with mock data)

* metric baseline support

* automated event layouting in rootcause chart

### CODE REVIEW

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js [hunk 1]
```diff
@@ -172,21 +176,23 @@ export default Ember.Controller.extend({
...
 
-    // NOTE: potential race condition?
-
     // update pending requests
     pending.delete(framework);
 
     // timeseries eviction
     // TODO optimize for same time range reload
-    incoming.forEach(e => delete timeseriesCache[e.urn]);
+    Object.keys(incoming).forEach(urn => delete timeseriesCache[urn]);
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: fyi, we can use a Map https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map
which has the 2 following methods: 

#### Map.prototype.clear()
```
Removes all key/value pairs from the Map object.
```
#### Map.prototype.delete(key)
```
Returns true if an element in the Map object existed and has been removed, or false if the element does not exist. Map.prototype.has(key) will return false afterwards.
```

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js [hunk 2]
```diff
@@ -246,43 +273,51 @@ export default Ember.Controller.extend({
...
       .then(res => res.json())
-      .then(this._extractTimeseries)
-      .then(incoming => this._completeRequestMissingTimeseries(this, incoming));
+      .then(res => this._extractTimeseries(res))
+      .then(timeseries => this._convertMetricToBaseline(timeseries, baselineOffset))
+      .then(timeseries => this._completeRequestMissingTimeseries(this, timeseries));
 
   },
 
   _completeRequestMissingTimeseries(that, incoming) {
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: Also the use of arrow functions should let us avoid passing the context. In which case is this failing?

Comment (justYves) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: ^ This is the only blocking comment

File: thirdeye/thirdeye-frontend/app/pods/rootcause/template.hbs [hunk 3]
```diff
@@ -59,4 +59,9 @@ urns
   columns=eventTableColumns
   selectedUrns=selectedUrns
   onSelect=(action "tableOnSelect")
+}}
+
+<h3>This is the heatmap.</h3>
+{{rootcause-heatmap
+  mode='change'
```
Comment (justYves) [inline thirdeye/thirdeye-frontend/app/pods/rootcause/template.hbs]: fyi, we `generally` follow this convention:  use single quotes in JS, double quotes in the template.

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-heatmap/component.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+          values[n][v] = transformation(curr, base, currTotal, baseTotal);
+        });
+      });
+
+      return values;
+    }
+  ),
+
+  _makeTransformation(mode) {
+    if (mode == 'change') {
```
[the remark this answers is not in the document]
Comment (ttbach) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-heatmap/component.js]: I second this comment.

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js [hunk 5]
```diff
@@ -120,13 +122,55 @@ export default Ember.Component.extend({
     }
   ),
 
+  _eventValues: Ember.computed(
```
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js]: fixed

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js [hunk 6]
```diff
@@ -120,13 +122,55 @@ export default Ember.Component.extend({
...
+          max = i > max ? i : max;
+
+        } else {
+          // remove
+          delete lanes[urn2lane[urn]];
+
+        }
+      });
+
+      const normalized = {};
```
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js]: hmm. that will rip it out of context

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js [hunk 7]
```diff
@@ -120,13 +122,55 @@ export default Ember.Component.extend({
...
+  ),
+
   _filterDisplayable(urns) {
     const { entities, timeseries } = this.getProperties('entities', 'timeseries');
 
     return [...urns]
       .filter(urn => entities[urn])
-      .filter(urn => ['metric', 'event'].includes(entities[urn].type))
-      .filter(urn => entities[urn].type != 'metric' || timeseries[urn]);
+      .filter(urn => ['metric', 'event', 'frontend:baseline:metric'].includes(entities[urn].type))
```
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js]: fixed

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js [hunk 8]
```diff
@@ -144,15 +188,27 @@ export default Ember.Component.extend({
...
+      const { timeseries } = this.getProperties('timeseries');
+
+      return {
+        timestamps: timeseries[entity.urn].timestamps,
+        values: timeseries[entity.urn].values,
+        type: 'scatter',
+        axis: 'y'
+      };
+
     } else if (entity.type == 'dimension') {
```
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-chart/component.js]: no. the entity type (not eventType!) carries semantic information for the chart. E.g. a metric is displayed differently than an event. Every type requires different logic to convert it into series data.

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-heatmap/component.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-heatmap/component.js]: fixed

File: thirdeye/thirdeye-frontend/app/pods/components/rootcause-heatmap/template.hbs [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+        {{#each-in dimValues as |dimValue value|}}
+          | {{dimValue}} ({{value}})
+        {{else}}
+          | No values
+        {{/each-in}}
+    </li>
+  {{else}}
+    <li>No dimensions</li>
+  {{/each-in}}
+</ol>
```
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/components/rootcause-heatmap/template.hbs]: fixed

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js [hunk 10]
```diff
@@ -162,6 +165,7 @@ export default Ember.Controller.extend({
       const url = this._makeFrameworkUrl(framework, context);
       fetch(url)
         .then(res => res.json())
```
[the remark this answers is not in the document]
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: fixed

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js (hunk 1 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: yeah I saw that one too. if we introduce it, we should replace all plain JS objects with it for consistency

File: thirdeye/thirdeye-frontend/app/pods/rootcause/template.hbs (hunk 3 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/template.hbs]: fixed

File: thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js (hunk 2 above)
Comment (apucher) [reply thirdeye/thirdeye-frontend/app/pods/rootcause/controller.js]: semi-fix. arrow context works for entities loading, but not for timeseries loading. this is a strange case.

**Pull request merged**